eos-paygd/main.c

```diff
@@ -24,14 +24,62 @@
 #include <libgsystemservice/service.h>
 #include <libeos-payg/service.h>
 #include <signal.h>
+#include <linux/reboot.h>
+#include <sys/reboot.h>
 
+/* Force a reboot in situations where we are not able to enforce PAYG */
+static gboolean
+sync_and_reboot (gpointer user_data)
+{
+    g_warning ("Forcing reboot now!");
+    sync ();
+    reboot (LINUX_REBOOT_CMD_RESTART);
+    return FALSE;
+}
 
 int
 main (int   argc,
       char *argv[])
 {
   g_autoptr(GError) error = NULL;
   g_autoptr(EpgService) service = NULL;
+  guint timeout_id;
+
+  /* If eos-paygd is running from the initramfs, change the process name so
+   * that it survives the pivot to the final root filesystem. This is an
+   * (ab)use of the functionality systemd has to support storage daemons needed
+   * for mounting/unmounting the root filesystem. See
+   * https://phabricator.endlessm.com/T27037
+   */
+  if (access ("/etc/initrd-release", F_OK) >= 0)
+    argv[0][0] = '@';
+  else
+    g_debug ("eos-paygd unexpectedly running from root filesystem!");
```

Comment:
It might even make sense to make this fatal - as in shutdown/reboot the PC fatal - but it should certainly scream as loudly in the logs as possible.

if we're not launching from initramfs, eos-paygd should be assumed to be compromised.  But we can't log the error in a way that lets the user know that it's compromised.  sigh.

I guess there's benefit to letting it run outside of initramfs for debug purposes, but I wonder if that should have a special compile time debug flag that only devs can use...

Reply:
So I added a meson option `debug_build` which lets eos-paygd run from the root fs, and documented it in the README. Without that, we emit a `g_warning` and initiate a reboot. But I think this means we should remove the integration tests in `eos-paygd/tests/eos-paygd.py`, for the following reasons:
- it would be very easy for a developer to omit the aforementioned meson option, run the tests, and have their computer reboot on them.
- the only thing they're testing is that eos-paygd **can't** be run as root which is no longer true. We could change the test to test that eos-paygd **can** be run as root but then we would also have to fool eos-paygd into thinking it was started from the initramfs (and it would not be very nice to create `/etc/initrd-release`, even temporarily in the test)

Reply:
>     * it would be very easy for a developer to omit the aforementioned meson option, run the tests, and have their computer reboot on them.

You could always `g_test_skip()` that test if the project has been built with `debug_build`.

Reply:
I have general concerns that as soon as one part of a project gets marked as “unfortunately not testable”, then there’s a slippery slope to other parts of the project getting lumped in with it as ‘not testable’, and then you never end up testing any of the hard bits of the project.

Reply:
> > ```
> > * it would be very easy for a developer to omit the aforementioned meson option, run the tests, and have their computer reboot on them.
> > ```
> 
> You could always `g_test_skip()` that test if the project has been built with `debug_build`.

Oh I guess you meant "if the project has **not** been built with `debug_build`". Hmm, would have to somehow check a C macro from python.